core/src/main/java/org/eclipse/hono/util/TenantObject.java

```diff
@@ -117,173 +122,103 @@ public TenantObject setEnabled(final boolean flag) {
     }
 
     /**
-     * Gets the subject DN of this tenant's configured trusted
-     * certificate authority.
-     * 
-     * @return The DN or {@code null} if no CA has been set.
-     */
-    @JsonIgnore
-    public X500Principal getTrustedCaSubjectDn() {
-
-        final JsonObject trustedCa = getProperty(TenantConstants.FIELD_PAYLOAD_TRUSTED_CA, JsonObject.class);
-        if (trustedCa == null) {
-            return null;
-        } else {
-            return Optional
-                    .ofNullable(getProperty(trustedCa, TenantConstants.FIELD_PAYLOAD_SUBJECT_DN, String.class))
-                    .map(dn -> new X500Principal(dn)).orElse(null);
-        }
-    }
-
-    /**
-     * Sets the trusted certificate authority to use for authenticating
+     * Adds the trusted certificate authority to use for authenticating
      * devices of this tenant.
      * 
-     * @param publicKey The CA's public key.
-     * @param subjectDn The CA's subject DN.
+     * @param publicKey The trusted CA's public key.
+     * @param subjectDn The trusted CA's subject DN.
+     * @param notBefore The start date in which the trustedCa becomes valid.
+     * @param notAfter The end date in which the trustedCa becomes invalid.
+     * 
      * @return This tenant for command chaining.
      * @throws NullPointerException if any of the parameters is {@code null}.
      */
     @JsonIgnore
-    public TenantObject setTrustAnchor(final PublicKey publicKey, final X500Principal subjectDn) {
+    public TenantObject addTrustAnchor(final PublicKey publicKey, final X500Principal subjectDn,
+            final Instant notBefore, final Instant notAfter) {
 
         Objects.requireNonNull(publicKey);
         Objects.requireNonNull(subjectDn);
+        Objects.requireNonNull(notBefore);
+        Objects.requireNonNull(notAfter);
 
         final JsonObject trustedCa = new JsonObject();
         trustedCa.put(TenantConstants.FIELD_PAYLOAD_SUBJECT_DN, subjectDn.getName(X500Principal.RFC2253));
         trustedCa.put(TenantConstants.FIELD_PAYLOAD_PUBLIC_KEY, publicKey.getEncoded());
+        trustedCa.put(TenantConstants.FIELD_PAYLOAD_NOT_BEFORE, notBefore.toEpochMilli());
```

Comment:
I think we will have a problem here. The device registry implementation currently uses `TenantObject` as its data format for storing tenant information in-memory. However, it uses `Tenant` and `TenantCertificateAuthority` from the service-base module in the Management API. `Tenant` instances are therefore mapped hence and forth to `TenantObject` instances.

The `TenantObject` however, has been originally designed as the value object for the result of the *get Tenant* operation of the Tenant API. The result does not contain *not-before* and *not-after* (because the Tenant service is supposed to have returned only currently valid CAs). 

If we want to continue to use `TenantObject` within the device registry as the data store format, then the *not-before* and *not-after* property format needs to match the format used in the `TenantCertificateAuthority`, which is the ISO string format ...

However, I am not really convinced that it is a good idea to use `TenantObject` as the internal data store format in the device registry because it requires us to keep the validity period properties in `TenantObject`, which the Tenant API doesn't require ... Instead, the device registry should probably go back to keeping the data in plain `JsonObject`s as we originally did ...

Reply:
> then the not-before and not-after property format needs to match the format used in the TenantCertificateAuthority, which is the ISO string format ...

this was the reason i used the `toString()` representation of the instant since the management API assumes that the format is an ISO format

> Instead, the device registry should probably go back to keeping the data in plain JsonObjects as we originally did ...

i have the feeling even we use a `JsonObject` to store the registry data in-memory, it will still contain these properties (since they are stored in a database/record system). if the tenantobject does not provide any API to retrieve the `not_before`/`not-after` properties, wouldn't that be what the Tenant API requires i.e hiding these properties from clients?

Reply:
@sophokles73 so whats the status of this PR? if you think we should definitely deprecate the `TenantObject` class and replace it with a `JsonObject` then i will suggest that we do that in a separate PR. this one is already quiet fat IMHO :-) WDYT?

Reply:
The device registry will in any case need to keep track of the *not-before* and *not-after* properties. The registry needs these properties to determine which CAs to return to an adapter.

The `TenantObject` class has been designed to contain all information defined by the response payload format of the Tenant API's *get Tenant* operation. It will continue to serve this purpose in the Hono Client.

The question is, if the `TenantObject` is also suitable as the data container format for the Device Registry. If it should serve this purpose as well, then it obviously needs to contain the *not-before* and *not-after* properties as well, and in that case it needs to keep them in the ISO String format.

However, if the registry contains all data as plain `JsonObject`s which comply with the schema defined by the Management API, then the `TenantObject` does not need to contain the validity period, because instances of it will only contain CAs which are valid at the moment the *get Tenant* request gets processed.

So we have two options, either 
1. keep the validity period in `TenantObject` and change the response payload of the *get Tenant* operation to also (optionally) include the *not-before* and *not-after* properties, or
2. change the Device Registry implementation to use plain `JsonObject`s for keeping the data in memory and adapt the mapping to/from the Management API accordingly

Thinking a little about it, I guess option 1 is easier to implement while still leaving us the option to later change the Device Registry implementation and no longer include the validity period in the response to *get Tenant*.

WDYT?

Reply:
@sophokles73 yes, I agree! I will go ahead and implement option one now.. 

> keep the validity period in TenantObject and change the response payload of the get Tenant operation to also (optionally) include the not-before and not-after properties, or

so this means that we should also update the Tenant API spec to (optionally) include these properties, right?

Reply:
exactly

Reply:
@sophokles73 i pushed the changes

Reply:
I thought we agreed that the *not-before* and the *not-after* should be stored in ISO format, didn't we?